### PR TITLE
boards: Add -fno-common to ARCHCFLAGS and ARCHCXXFLAGS 

### PR DESCRIPTION
**Pull request by xiaoxiang781216**
## Summary
Follow up the change: https://github.com/apache/incubator-nuttx/pull/5485

## Impact
elf loader

## Testing
Pass CI

### CODE REVIEW

**Comment by Ouss4** [comment]
@xiaoxiang781216 this also depends on https://github.com/espressif/esp-wireless-drivers-3rdparty/pull/1 right?

**Comment by xiaoxiang781216** [comment]
> @xiaoxiang781216 this also depends on [espressif/esp-wireless-drivers-3rdparty#1](https://github.com/espressif/esp-wireless-drivers-3rdparty/pull/1) right?

I remove esp32's Make.defs from this patch. Once the above patch get merged, I will send the new patch.

File: boards/arm/stm32f7/stm32f746g-disco/scripts/Make.defs [hunk 1]
```diff
@@ -30,14 +30,14 @@ ifneq ($(CONFIG_DEBUG_NOOPT),y)
...
+ARCHCXXFLAGS = -fno-common -fno-builtin -fno-exceptions -fcheck-new -fno-rtti
 
 ifeq ($(CONFIG_ARMV7M_TOOLCHAIN_CLANGL),y)
-  ARCHCFLAGS += -nostdlib -ffreestanding
-  ARCHCXXFLAGS += -nostdlib -ffreestanding
+  ARCHCFLAGS += -fno-common -nostdlib -ffreestanding
+  ARCHCXXFLAGS += -fno-common -nostdlib -ffreestanding
 else
-  ARCHCXXFLAGS += -fno-rtti
+  ARCHCXXFLAGS += -fno-common -fno-rtti
```
Comment (pkarashchenko) [inline boards/arm/stm32f7/stm32f746g-disco/scripts/Make.defs:40]: It's strange that `-fno-rtti` is here conditionally and at line 34 unconditionally.

File: boards/arm/kinetis/teensy-3.x/src/teensy-3x.h [hunk 2]
```diff
@@ -55,16 +55,6 @@
...
 
-/****************************************************************************
- * Public Types
- ****************************************************************************/
-
-/****************************************************************************
- * Public Data
- ****************************************************************************/
-
-FAR struct i2c_master_s *i2c_dev;
```
Comment (pkarashchenko) [inline boards/arm/kinetis/teensy-3.x/src/teensy-3x.h:66]: Great finding!

File: boards/arm/kinetis/teensy-3.x/src/k20_i2c.c [hunk 3]
```diff
@@ -53,7 +53,7 @@
 
 void kinetis_i2cdev_initialize(void)
 {
-  i2c_dev = NULL;
+  FAR struct i2c_master_s *i2c_dev = NULL;
```
Comment (pkarashchenko) [inline boards/arm/kinetis/teensy-3.x/src/k20_i2c.c:56]: Great finding!

File: boards/arm/sam34/sam4cmp-db/scripts/Make.defs [hunk 4]
```diff
@@ -32,8 +32,8 @@ ifneq ($(CONFIG_DEBUG_NOOPT),y)
   ARCHOPTIMIZATION += $(MAXOPTIMIZATION) -fno-strict-aliasing
 endif
 
-ARCHCFLAGS = -fno-builtin
-ARCHCXXFLAGS = -fno-builtin -fno-exceptions -fcheck-new
+ARCHCFLAGS = -fno-common -fno-builtin
+ARCHCXXFLAGS = -fno-common -fno-builtin -fno-exceptions -fcheck-new
```
Comment (pkarashchenko) [inline boards/arm/sam34/sam4cmp-db/scripts/Make.defs:36]: Note: strange that `boards/arm/sam34/sam3u-ek/scripts/Make.defs` has `-fno-rtti` and here we do not have it. The same in other files. This is not related to this PR, but notice in general. Maybe we should have a separate PR to unify `-fno-rtti` usage

File: boards/xtensa/esp32s3/esp32s3-devkit/scripts/Make.defs [hunk 5]
```diff
@@ -36,8 +36,8 @@ ifneq ($(CONFIG_DEBUG_NOOPT),y)
   ARCHOPTIMIZATION += $(MAXOPTIMIZATION) -fno-strict-aliasing -fno-strength-reduce
 endif
 
-ARCHCFLAGS = -fno-builtin -ffunction-sections -fdata-sections -fstrict-volatile-bitfields -mlongcalls
-ARCHCXXFLAGS = $(ARCHCFLAGS) -fno-exceptions -fcheck-new -fno-rtti
+ARCHCFLAGS = -fno-common -fno-builtin -ffunction-sections -fdata-sections -fstrict-volatile-bitfields -mlongcalls
+ARCHCXXFLAGS = -fno-common $(ARCHCFLAGS) -fno-exceptions -fcheck-new -fno-rtti
```
Comment (pkarashchenko) [inline boards/xtensa/esp32s3/esp32s3-devkit/scripts/Make.defs:40]:
```suggestion
ARCHCXXFLAGS = $(ARCHCFLAGS) -fno-exceptions -fcheck-new -fno-rtti
```

File: boards/xtensa/esp32s2/esp32s2-saola-1/scripts/Make.defs [hunk 6]
```diff
@@ -46,16 +46,16 @@ ifneq ($(CONFIG_DEBUG_NOOPT),y)
...
+ARCHCXXFLAGS = -fno-common $(ARCHCFLAGS) -fno-exceptions -fcheck-new -fno-rtti
 ARCHWARNINGS = -Wall -Wstrict-prototypes -Wshadow -Wundef
 ARCHWARNINGSXX = -Wall -Wshadow -Wundef
 ARCHPICFLAGS = -fpic
 
 # if SPIRAM/PSRAM is used then we need to include a workaround
 
 ifeq ($(CONFIG_ESP32S2_SPIRAM),y)
-  ARCHCFLAGS += -mfix-esp32s2-psram-cache-issue
+  ARCHCFLAGS += -fno-common -mfix-esp32s2-psram-cache-issue
```
Comment (pkarashchenko) [inline boards/xtensa/esp32s2/esp32s2-saola-1/scripts/Make.defs:58]:
```suggestion
  ARCHCFLAGS += -mfix-esp32s2-psram-cache-issue
```

File: boards/risc-v/bl602/bl602evb/scripts/Make.defs [hunk 7]
```diff
@@ -48,7 +48,7 @@ ARCHWARNINGSXX = -Wall -Wshadow -Wundef
 ARCHPICFLAGS = -fpic -msingle-pic-base -mpic-register=r10
 
 ifeq ($(CONFIG_STACK_OVERFLOW_CHECK),y)
-  ARCHCFLAGS += -finstrument-functions -ffixed-s11
+  ARCHCFLAGS += -fno-common -finstrument-functions -ffixed-s11
```
Comment (pkarashchenko) [inline boards/risc-v/bl602/bl602evb/scripts/Make.defs:51]:
```suggestion
  ARCHCFLAGS += -finstrument-functions -ffixed-s11
```

File: boards/mips/pic32mz/pic32mz-starterkit/scripts/Make.defs [hunk 8]
```diff
@@ -59,14 +59,14 @@ ifneq ($(CONFIG_DEBUG_NOOPT),y)
...
-ARCHCXXFLAGS = -fno-builtin -fno-exceptions -fcheck-new
+ARCHCFLAGS = -fno-common -fno-builtin
+ARCHCXXFLAGS = -fno-common -fno-builtin -fno-exceptions -fcheck-new
 ARCHWARNINGS = -Wall -Wstrict-prototypes -Wshadow -Wundef
 ARCHWARNINGSXX = -Wall -Wshadow -Wundef
 
 ifeq ($(CONFIG_MIPS32_TOOLCHAIN_MICROCHIPL_XC32),y)
 ifeq ($(CONFIG_DEBUG_FEATURES),y)
-  ARCHCFLAGS += -D__DEBUG -D__MPLAB_DEBUGGER_PK3=1 -fframe-base-loclist
+  ARCHCFLAGS += -fno-common -D__DEBUG -D__MPLAB_DEBUGGER_PK3=1 -fframe-base-loclist
```
Comment (pkarashchenko) [inline boards/mips/pic32mz/pic32mz-starterkit/scripts/Make.defs:69]:
```suggestion
  ARCHCFLAGS += -D__DEBUG -D__MPLAB_DEBUGGER_PK3=1 -fframe-base-loclist
```

File: boards/mips/pic32mz/flipnclick-pic32mz/scripts/Make.defs [hunk 9]
```diff
@@ -59,14 +59,14 @@ ifneq ($(CONFIG_DEBUG_NOOPT),y)
...
-ARCHCXXFLAGS = -fno-builtin -fno-exceptions -fcheck-new
+ARCHCFLAGS = -fno-common -fno-builtin
+ARCHCXXFLAGS = -fno-common -fno-builtin -fno-exceptions -fcheck-new
 ARCHWARNINGS = -Wall -Wstrict-prototypes -Wshadow -Wundef
 ARCHWARNINGSXX = -Wall -Wshadow -Wundef
 
 ifeq ($(CONFIG_MIPS32_TOOLCHAIN_MICROCHIPL_XC32),y)
 ifeq ($(CONFIG_DEBUG_FEATURES),y)
-  ARCHCFLAGS += -D__DEBUG -D__MPLAB_DEBUGGER_PK3=1 -fframe-base-loclist
+  ARCHCFLAGS += -fno-common -D__DEBUG -D__MPLAB_DEBUGGER_PK3=1 -fframe-base-loclist
```
Comment (pkarashchenko) [inline boards/mips/pic32mz/flipnclick-pic32mz/scripts/Make.defs:69]:
```suggestion
  ARCHCFLAGS += -D__DEBUG -D__MPLAB_DEBUGGER_PK3=1 -fframe-base-loclist
```

File: boards/mips/pic32mz/chipkit-wifire/scripts/Make.defs [hunk 10]
```diff
@@ -59,14 +59,14 @@ ifneq ($(CONFIG_DEBUG_NOOPT),y)
...
-ARCHCXXFLAGS = -fno-builtin -fno-exceptions -fcheck-new
+ARCHCFLAGS = -fno-common -fno-builtin
+ARCHCXXFLAGS = -fno-common -fno-builtin -fno-exceptions -fcheck-new
 ARCHWARNINGS = -Wall -Wstrict-prototypes -Wshadow -Wundef
 ARCHWARNINGSXX = -Wall -Wshadow -Wundef
 
 ifeq ($(CONFIG_MIPS32_TOOLCHAIN_MICROCHIPL_XC32),y)
 ifeq ($(CONFIG_DEBUG_FEATURES),y)
-  ARCHCFLAGS += -D__DEBUG -D__MPLAB_DEBUGGER_PK3=1 -fframe-base-loclist
+  ARCHCFLAGS += -fno-common -D__DEBUG -D__MPLAB_DEBUGGER_PK3=1 -fframe-base-loclist
```
Comment (pkarashchenko) [inline boards/mips/pic32mz/chipkit-wifire/scripts/Make.defs:69]:
```suggestion
  ARCHCFLAGS += -D__DEBUG -D__MPLAB_DEBUGGER_PK3=1 -fframe-base-loclist
```

File: boards/arm/stm32/stm32f4discovery/scripts/Make.defs [hunk 11]
```diff
@@ -34,18 +34,18 @@ ifneq ($(CONFIG_DEBUG_NOOPT),y)
...
+ARCHCXXFLAGS = -fno-common -fno-builtin -fno-exceptions -fcheck-new
 ARCHWARNINGS = -Wall -Wstrict-prototypes -Wshadow -Wundef
 ARCHWARNINGSXX = -Wall -Wshadow -Wundef
 ARCHPICFLAGS = -fpic -msingle-pic-base -mpic-register=r10
 
 ifeq ($(CONFIG_ARMV7M_TOOLCHAIN_CLANGL),y)
-  ARCHCFLAGS += -nostdlib -ffreestanding
-  ARCHCXXFLAGS += -nostdlib -ffreestanding
+  ARCHCFLAGS += -fno-common -nostdlib -ffreestanding
+  ARCHCXXFLAGS += -fno-common -nostdlib -ffreestanding
```
Comment (pkarashchenko) [inline boards/arm/stm32/stm32f4discovery/scripts/Make.defs:45]:
```suggestion
  ARCHCFLAGS += -nostdlib -ffreestanding
  ARCHCXXFLAGS += -nostdlib -ffreestanding
```

File: boards/arm/stm32/stm32f4discovery/scripts/Make.defs [hunk 12]
```diff
@@ -34,18 +34,18 @@ ifneq ($(CONFIG_DEBUG_NOOPT),y)
...
 else
-  ARCHCFLAGS += -funwind-tables
-  ARCHCXXFLAGS += -fno-rtti -funwind-tables
+  ARCHCFLAGS += -fno-common -funwind-tables
+  ARCHCXXFLAGS += -fno-common -fno-rtti -funwind-tables
```
Comment (pkarashchenko) [inline boards/arm/stm32/stm32f4discovery/scripts/Make.defs:48]:
```suggestion
  ARCHCFLAGS += -funwind-tables
  ARCHCXXFLAGS += -fno-rtti -funwind-tables
```

File: boards/arm/stm32/stm32f4discovery/configs/testlibcxx/Make.defs [hunk 13]
```diff
@@ -34,7 +34,7 @@ ifneq ($(CONFIG_DEBUG_NOOPT),y)
   ARCHOPTIMIZATION += $(MAXOPTIMIZATION) -fno-strict-aliasing
 endif
 
-ARCHCFLAGS = -fno-builtin
+ARCHCFLAGS = -fno-common -fno-builtin
 ifeq ($(CONFIG_CXX_EXCEPTION),y)
   ARCHCPUFLAGSXX = -fno-builtin
```
Comment (pkarashchenko) [inline boards/arm/stm32/stm32f4discovery/configs/testlibcxx/Make.defs:39]:
```suggestion
  ARCHCPUFLAGSXX = -fno-common -fno-builtin
```
and same for `else` case

File: boards/arm/stm32/stm32f4discovery/configs/cxxtest/Make.defs [hunk 14]
```diff
@@ -33,7 +33,7 @@ ifneq ($(CONFIG_DEBUG_NOOPT),y)
   ARCHOPTIMIZATION += $(MAXOPTIMIZATION) -fno-strict-aliasing
 endif
 
-ARCHCFLAGS = -fno-builtin
+ARCHCFLAGS = -fno-common -fno-builtin
 ifeq ($(CONFIG_CXX_EXCEPTION),y)
   ARCHCPUFLAGSXX = -fno-builtin
```
Comment (pkarashchenko) [inline boards/arm/stm32/stm32f4discovery/configs/cxxtest/Make.defs:38]:
```suggestion
  ARCHCPUFLAGSXX = -fno-common -fno-builtin
```
and same for `else` case

File: boards/arm/stm32/nucleo-f446re/scripts/Make.defs [hunk 15]
```diff
@@ -35,14 +35,14 @@ ifneq ($(CONFIG_DEBUG_NOOPT),y)
...
+ARCHCXXFLAGS = -fno-common -fno-builtin -fno-exceptions -fcheck-new
 
 ifeq ($(CONFIG_ARMV7M_TOOLCHAIN_CLANGL),y)
-  ARCHCFLAGS += -nostdlib -ffreestanding
-  ARCHCXXFLAGS += -nostdlib -ffreestanding
+  ARCHCFLAGS += -fno-common -nostdlib -ffreestanding
+  ARCHCXXFLAGS += -fno-common -nostdlib -ffreestanding
 else
-  ARCHCXXFLAGS += -fno-rtti
+  ARCHCXXFLAGS += -fno-common -fno-rtti
```
Comment (pkarashchenko) [inline boards/arm/stm32/nucleo-f446re/scripts/Make.defs:45]:
```suggestion
  ARCHCXXFLAGS += -fno-rtti
```

File: boards/arm/stm32/nucleo-f446re/scripts/Make.defs [hunk 16]
```diff
@@ -35,14 +35,14 @@ ifneq ($(CONFIG_DEBUG_NOOPT),y)
...
-ARCHCFLAGS = -fno-builtin
-ARCHCXXFLAGS = -fno-builtin -fno-exceptions -fcheck-new
+ARCHCFLAGS = -fno-common -fno-builtin
+ARCHCXXFLAGS = -fno-common -fno-builtin -fno-exceptions -fcheck-new
 
 ifeq ($(CONFIG_ARMV7M_TOOLCHAIN_CLANGL),y)
-  ARCHCFLAGS += -nostdlib -ffreestanding
-  ARCHCXXFLAGS += -nostdlib -ffreestanding
+  ARCHCFLAGS += -fno-common -nostdlib -ffreestanding
+  ARCHCXXFLAGS += -fno-common -nostdlib -ffreestanding
```
Comment (pkarashchenko) [inline boards/arm/stm32/nucleo-f446re/scripts/Make.defs:43]:
```suggestion
  ARCHCFLAGS += -nostdlib -ffreestanding
  ARCHCXXFLAGS += -nostdlib -ffreestanding
```

File: boards/arm/lpc43xx/bambino-200e/configs/netnsh/Make.defs [hunk 17]
```diff
@@ -50,7 +50,7 @@ ifneq ($(CONFIG_DEBUG_NOOPT),y)
   ARCHOPTIMIZATION += $(MAXOPTIMIZATION) -fno-strict-aliasing
 endif
 
-ARCHCFLAGS = -fno-builtin
+ARCHCFLAGS = -fno-common -fno-builtin
 ifeq ($(CONFIG_CXX_EXCEPTION),y)
   ARCHCPUFLAGSXX = -fno-builtin
```
Comment (pkarashchenko) [inline boards/arm/lpc43xx/bambino-200e/configs/netnsh/Make.defs:55]:
```suggestion
  ARCHCPUFLAGSXX = -fno-common -fno-builtin
```
and for `else` case

File: boards/arm/stm32f7/stm32f746g-disco/scripts/Make.defs (hunk 1 above)
Comment (xiaoxiang781216) [reply boards/arm/stm32f7/stm32f746g-disco/scripts/Make.defs:40]: Fixed

File: boards/xtensa/esp32s3/esp32s3-devkit/scripts/Make.defs (hunk 5 above)
Comment (xiaoxiang781216) [reply boards/xtensa/esp32s3/esp32s3-devkit/scripts/Make.defs:40]: Fixed

File: boards/risc-v/bl602/bl602evb/scripts/Make.defs (hunk 7 above)
Comment (xiaoxiang781216) [reply boards/risc-v/bl602/bl602evb/scripts/Make.defs:51]: Fixed

File: boards/mips/pic32mz/flipnclick-pic32mz/scripts/Make.defs (hunk 9 above)
Comment (xiaoxiang781216) [reply boards/mips/pic32mz/flipnclick-pic32mz/scripts/Make.defs:69]: Done.

File: boards/mips/pic32mz/pic32mz-starterkit/scripts/Make.defs (hunk 8 above)
Comment (xiaoxiang781216) [reply boards/mips/pic32mz/pic32mz-starterkit/scripts/Make.defs:69]: Done.

File: boards/mips/pic32mz/chipkit-wifire/scripts/Make.defs (hunk 10 above)
Comment (xiaoxiang781216) [reply boards/mips/pic32mz/chipkit-wifire/scripts/Make.defs:69]: Done.

File: boards/arm/stm32/stm32f4discovery/scripts/Make.defs (hunk 11 above)
Comment (xiaoxiang781216) [reply boards/arm/stm32/stm32f4discovery/scripts/Make.defs:45]: Done.

File: boards/arm/stm32/stm32f4discovery/scripts/Make.defs (hunk 12 above)
Comment (xiaoxiang781216) [reply boards/arm/stm32/stm32f4discovery/scripts/Make.defs:48]: Done.

File: boards/arm/lpc43xx/bambino-200e/configs/netnsh/Make.defs (hunk 17 above)
Comment (xiaoxiang781216) [reply boards/arm/lpc43xx/bambino-200e/configs/netnsh/Make.defs:55]: Remove from https://github.com/apache/incubator-nuttx/pull/5555

File: boards/arm/stm32/nucleo-f446re/scripts/Make.defs (hunk 16 above)
Comment (xiaoxiang781216) [reply boards/arm/stm32/nucleo-f446re/scripts/Make.defs:43]: Remove from https://github.com/apache/incubator-nuttx/pull/5555

File: boards/arm/stm32/stm32f4discovery/configs/cxxtest/Make.defs (hunk 14 above)
Comment (xiaoxiang781216) [reply boards/arm/stm32/stm32f4discovery/configs/cxxtest/Make.defs:38]: Remove from https://github.com/apache/incubator-nuttx/pull/5555

File: boards/arm/stm32/stm32f4discovery/configs/testlibcxx/Make.defs (hunk 13 above)
Comment (xiaoxiang781216) [reply boards/arm/stm32/stm32f4discovery/configs/testlibcxx/Make.defs:39]: Remove from https://github.com/apache/incubator-nuttx/pull/5555

File: boards/arm/sam34/sam4cmp-db/scripts/Make.defs (hunk 4 above)
Comment (xiaoxiang781216) [reply boards/arm/sam34/sam4cmp-db/scripts/Make.defs:36]: Agree.

File: boards/xtensa/esp32s2/esp32s2-saola-1/scripts/Make.defs (hunk 6 above)
Comment (xiaoxiang781216) [reply boards/xtensa/esp32s2/esp32s2-saola-1/scripts/Make.defs:58]: Done.

File: boards/arm/stm32/nucleo-f446re/scripts/Make.defs (hunk 15 above)
Comment (xiaoxiang781216) [reply boards/arm/stm32/nucleo-f446re/scripts/Make.defs:45]: Done.

File: boards/arm/a1x/pcduino-a10/scripts/Make.defs [hunk 18]
```diff
@@ -35,8 +35,8 @@ ifneq ($(CONFIG_DEBUG_NOOPT),y)
 endif
 
 ARCHCPUFLAGS = -mcpu=cortex-a8 -mfpu=vfpv4-d16
-ARCHCFLAGS = -fno-builtin
-ARCHCXXFLAGS = -fno-builtin -fno-exceptions -fcheck-new -fno-rtti
+ARCHCFLAGS = -fno-common -fno-builtin
+ARCHCXXFLAGS = -fno-common -fno-builtin -fno-exceptions -fcheck-new -fno-rtti
```
Comment (pkarashchenko) [inline boards/arm/a1x/pcduino-a10/scripts/Make.defs:39]: I think we should migrate this to `ARCHCXXFLAGS = $(ARCHCFLAGS) ...` in the next PR if possible

**Review by pkarashchenko — APPROVED** [review]

File: boards/arm/a1x/pcduino-a10/scripts/Make.defs (hunk 18 above)
Comment (xiaoxiang781216) [reply boards/arm/a1x/pcduino-a10/scripts/Make.defs:39]: Yes, I think so.

**Comment by pkarashchenko** [comment]
@Ouss4 @masayuki2009 could you please approve the changes?

**Pull request merged**